rfcs/20210510-binary-size-reduction/README.md

```diff
@@ -0,0 +1,209 @@
+* Rest functionality takes ~12 MB + ~2 MB of core API.
+
+### Updated disk space chart
+Current breakdown for future v2.4 release for disk space of certain parts:
+
+* ~2 MB for core API.
+* ~12 MB for core compute functionality: convolution, deconvolution, inner
+  product and matmul.
+* ~14 MB is GEMM auto-generated kernels (would be part of compute unless
+  replaced)
+* ~11 MB are reorders (~10 MB are instances of simple reorders which could be
+  replaced with jit-ted one)
+* ~11 MB is the rest functionality.
+
+## Background
+The reason why static linking can't solve the problem itself is the way library
+API is implemented. The call to `dnnl_primitive_desc_create` (which is also used
+underneath the correspondent C++ API) relies on statically defined
+implementation lists which contain all primitives at once. That means that if
+user calls the function (which is an essential part of the API), they get the
+full library in their application. And there are not many choices how to deal
+with such problem. One is to switch off implementations in compile time (this
+RFC proposal) or using a different API which would separate things on different
+levels (at least on primitive one).
+
+## Proposal
+To control the size, the proposal is to utilize the power of build system and
+let the user specify a set of desired functionality. The idea is to have sets of
+macros, one is external and would be defined by the build system based on user
+choices. This external set would control another set - internal - which lives
+inside the library and modifies implementation lists. Each primitive would get
+several macros to switch on/off certain implementations during compilation time.
+Further optimizations of disk space will be provided by linker which excludes
+unused code inside the end-user application unless certain scenario happens [2].
+
+The proposal is to introduce three new build options which would define the list
+of available implementations for user at build time:
+* DNNL_ENABLE_PROP_KIND. This option supports two values: TRAINING (the default)
+  and INFERENCE. Second one is a subset of first one and strips backward off.
+* DNNL_ENABLE_PRIMITIVE. This option supports all primitive names as they are
+  used in the library API. E.g. `-DDNNL_ENABLE_PRIMITIVE=CONVOLUTION;REORDER`.
+  Anything that is not listed explicitly will be excluded from lists.
+  Enumeration comes with semicolon as a standard CMake string delimiter.
+* DNNL_ENABLE_PRIMITIVE_ISA. This option supports ISA names as they are used in
+  the library API for `DNNL_SET_MAX_CPU_ISA`. E.g.
+  `-DDNNL_ENABLE_PRIMITIVE_ISA=AVX2;AVX512_CORE_AMX`. Anything that is not
+  specified explicitly will be excluded except compiler-based optimization
+  codes. This relates to "simple", "GEMM"-based and reference implementations.
+  Only direct ISA-related codes would be affected by this option, e.g.
+  `jit_avx2_convolution_fwd_t`.
```

Comment:
I wonder how we should define that one... In particular there are a bunch of situations where we have no dedicated impl for ISA X, but we dispatch dedicated impl for ISA Y with Y < X. For example what should happen when a user passes `AVX`? With the current proposal, I assume this will just include ref and gemm based impls, right? However a user might expect to still get some sse41 implementations I guess.

So I agree with the idea of having this build knob exposed, but I think we should polish its behavior to make sure it fits the user expectation.

Reply:
This option is specifically about enabling. By default `ALL` is enabled. If someone asks to enable `AVX` only, than all `AVX`-related implementations would be switched on. If `SSE41` is desired on top, then it should be asked in the option arguments as well.
In general, I think one doesn't want a fallback to SSE41 code since it would be slower and it's pretty rare case when something is not available for certain ISA when it comes not to modern ISA where most fall backs for, let's say, `f32` would happen into AVX512-CORE, like VNNI ISA.

Making user explicitly specify what they want removes the implicit assumptions from both sides and the build system gets a pretty clear contract what to do.

We may consider adding `DNNL_DISABLE_PRIMITIVE_ISA` which on contrary would only removed specified options from the list, but don't see this is needed for OV folks right now and it wouldn't be a big challenge to add such.

@mgouicem, does it help?

Reply:
Maybe my example was not good enough, but yours works as well :-)
If user enables only AMX, we have a bunch of primitives that do not have an AMX specific implementation. For those primitives, the user will likely not want to falls back to reference but whatever would be normally dispatched on an AMX system.

Reply:
My best guess that people who want AMX don't care about final binary size. Another thing with AMX is that it's the top and if people enables it, there's nothing to get rid of, maybe SSE41 and AVX2 but I don't expect the gain would be big from removing these two.

I would say this particular feature targets three cases:
1. Leave just SSE41 in the final binary.
2. Leave just AVX2 in the final binary.
3. I don't care about the size.

Not sure what else we should pursuit here when we didn't get any other feedback expect caring for first two options from OV side.

I can add a note for modern ISA relies on a fall back into AVX512-CORE at least so that users better include it. Will that suffice?

Reply:
Once again this is not specific to AMX. Even in the "leave AVX2" case, we might still have to dispatch non AVX2 implementations. For example, we could have fallbacks to sse41 or jit_uni, or brgemm impls.
I would expect that the users of that option would not want to see performance drop, just to eliminate undispatched implementations (e.g. get rid of avx512 if they will never run on avx512 machines).

@dmitry-gorokhov could you comment on this proposal?

Reply:
I think as first (and good enough for most of the cases) we can consider disabling only "higher" isa. Like avx512 for customers targeting avx2 only. 

Is also makes sense to consider conditional compilation by precision. There might be requests to build the product w/o int8 or bf16 support.

Reply:
> e.g. get rid of avx512 if they will never run on avx512 machines

+1 to this. User will probably want to disable implementations that will never be used due to unsupported ISA in their system.

Reply:
Though I'm not sure ISA changes can make it to 2.4. @dzarukin?

Reply:
ISA extension is removed from original RFC and would be granted later, targeting v2.5 with, possibly, other improvements.

Reply:
Sounds good to me, thanks.